Mathlib/NumberTheory/LSeries/Deriv.lean

```diff
@@ -0,0 +1,227 @@
+the problem that this expression does not type-check. Similarly, we make use of the abbreviation
+`LSeries.logPowMul n f` for `log^n * f`.
+-/
+
+-- Should this go into its own file (because it needs `EReal`s, it doesn't fit well
+-- in some existing file), together with versions for the left, upper and lower half-planes?
+/-- An open right half-plane is open in the complex plane. -/
+lemma Complex.isOpen_rightHalfPlane (x : EReal) : IsOpen {z : ℂ | x < z.re} :=
+  isOpen_lt continuous_const <| EReal.continuous_coe_iff.mpr continuous_re
+
+open Complex LSeries
+
+/-!
+### Multiplication by (powers of) log
+-/
+
+/-- The (point-wise) product of `log : ℕ → ℂ` with `f`. -/
+noncomputable abbrev LSeries.logMul (f : ℕ → ℂ) (n : ℕ) : ℂ := log n * f n
+
+
+/-- The (point-wise) product of the `m`th power of `log` with `f`. -/
+noncomputable abbrev logPowMul (m : ℕ) (f : ℕ → ℂ) (n : ℕ) : ℂ :=
+  (log n) ^ m * f n
+
+lemma logPowMul_zero (f : ℕ → ℂ) : logPowMul 0 f = f := by
+  ext
+  simp only [logPowMul, pow_zero, one_mul]
+
+lemma logPowMul_succ (m : ℕ) (f : ℕ → ℂ) : logPowMul m.succ f = logMul (logPowMul m f) := by
+  ext
+  simp only [logPowMul, pow_succ, mul_assoc, logMul]
+
+lemma logPowMul_succ' (m : ℕ) (f : ℕ → ℂ) : logPowMul m.succ f = logPowMul m (logMul f) := by
+  ext m
+  simp only [logPowMul, pow_succ, logMul, ← mul_assoc]
+  rw [mul_comm (log m)]
+
+/-!
+### The derivative of an L-series
+-/
+
+/-- The derivative of the terms of an L-series. -/
+lemma LSeries.hasDerivAt_term (f : ℕ → ℂ) (n : ℕ) (s : ℂ) :
+    HasDerivAt (fun z ↦ term f z n) (-(term (logMul f) s n)) s := by
+  rcases n.eq_zero_or_pos with rfl | hn
+  · simp only [term_zero, neg_zero, hasDerivAt_const]
+  simp only [term_of_ne_zero hn.ne']
+  rw [← neg_div, ← neg_mul, mul_comm, mul_div_assoc]
+  simp_rw [div_eq_mul_inv, ← cpow_neg]
+  refine HasDerivAt.const_mul (f n) ?_
+  rw [mul_comm, ← mul_neg_one (Complex.log n), ← mul_assoc]
+  exact (hasDerivAt_neg' s).const_cpow (Or.inl <| Nat.cast_ne_zero.mpr hn.ne')
+
+/-- The derivative of the terms of an L-series. -/
+lemma LSeries.deriv_term (f : ℕ → ℂ) (n : ℕ) (s : ℂ) :
+    deriv (fun z ↦ term f z n) s = -(term (logMul f) s n) :=
+  (hasDerivAt_term f n s).deriv
+
+/-- The derivative of the terms of an L-series. -/
+lemma LSeries.deriv_term' (f : ℕ → ℂ) (n : ℕ) :
+    deriv (fun z ↦ term f z n) = fun s ↦ -(term (logMul f) s n) :=
+  funext <| deriv_term f n
+
+private
+lemma LSeries.LSeriesSummable_logMul_and_hasDerivAt {f : ℕ → ℂ} {s : ℂ}
+    (h : abscissaOfAbsConv f < s.re) :
+    LSeriesSummable (logMul f) s ∧ HasDerivAt (LSeries f) (- LSeries (logMul f) s) s := by
+  -- The L-series of `f` is summable at some real `x < re s`.
+  obtain ⟨x, h', hf⟩ := LSeriesSummable_lt_re_of_abscissaOfAbsConv_lt_re h
+  -- We work in the right half-plane `re z > (x + re s)/2`, where we have a uniform
+  -- summable bound on `‖term f z ·‖`.
+  let S : Set ℂ := {z | (x + s.re) / 2 < z.re}
```

Comment:
I usually go for `exists_between` for this sort of thing rather than taking the explicit midpoint. I had a reviewer patch in the works which golfed this proof considerably (reducing h3 and h4 to one line each) but it conflicts with your latest commit, I'll try rebasing it.

Reply:
OK.